astropy/utils/xml/src/iterparse.c

```diff
@@ -384,25 +385,24 @@ startElement(IterParser *self, const XML_Char *name, const XML_Char **atts)
                 goto fail;
             }
             do {
-                if (*(*(att_ptr + 1)) != 0) {
-                    key = PyUnicode_FromString(*att_ptr);
-                    if (key == NULL) {
-                        goto fail;
-                    }
-                    val = PyUnicode_FromString(*(att_ptr + 1));
-                    if (val == NULL) {
-                        Py_DECREF(key);
-                        goto fail;
-                    }
-                    if (PyDict_SetItem(pyatts, key, val)) {
-                        Py_DECREF(key);
-                        Py_DECREF(val);
-                        goto fail;
-                    }
+                key = PyUnicode_FromString(*att_ptr);
+                if (key == NULL) {
+                    goto fail;
+                }
+                val = PyUnicode_FromString(*(att_ptr + 1));
```

Comment:
So it won't crash here without the `if (*(*(att_ptr + 1)) != 0)` check if there is no `(att_ptr + 1)`? Or is `(att_ptr + 1)` always guaranteed?

Reply:
I don't think this adds any new danger.  Since the `if` statement I removed also referenced that pointer,  it would have crashed there before if  `att_ptr + 1` were not a valid pointer:
`if (*(*(att_ptr + 1)) != 0)`

We could add a new check here, but I think there are multiple places in this loop that assume that the attribute structure delivered by expat was a fairly solid contract , having `att_ptr` be an array of an even number of valid strings (char *'s) followed by a `NULL` to end the loop.  So if `att_ptr` is not `NULL`, then `att_ptr` would also be not `NULL`.  So while we could check for `NULL`, we couldn't check for random garbage which might be just as likely if the expat attribute structure was corrupt.